File: apps/ios/Wewed/Services/TokenVerifier.swift
import Foundation
import CryptoKit

public struct ParsedQRToken: Equatable, Sendable {
    public let version: String
    public let weddingShortId: String
    public let passSerial: String
    public let eventBitmask: UInt8
    public let nonce: String
    public let signature: String
}

public enum TokenVerificationError: Error, Equatable, Sendable {
    case invalidFormat
    case unsupportedVersion
    case signatureMismatch
    case unauthorizedEvent
    case invalidKey
}

public enum TokenVerifier {
    public static func parse(token: String) -> Result<ParsedQRToken, TokenVerificationError> {
        let parts = token.split(separator: ".")
        guard parts.count == 6 else {
            return .failure(.invalidFormat)
        }

        let version = String(parts[0])
        guard version == "WW1" || version == "WW2" else {
            return .failure(.unsupportedVersion)
        }

        let weddingShortId = String(parts[1])
        let passSerial = String(parts[2])
        guard let bitmask = UInt8(parts[3], radix: 16) else {
            return .failure(.invalidFormat)
        }
        let nonce = String(parts[4])
        let signature = String(parts[5])

        return .success(ParsedQRToken(
            version: version,
            weddingShortId: weddingShortId,
            passSerial: passSerial,
            eventBitmask: bitmask,
            nonce: nonce,
            signature: signature
        ))
    }

    /// Asymmetric ECDSA (NIST P-256 / SHA-256) signature verification (WW2 Canonical Standard)
    public static func verifyAsymmetric(
        token: String,
        rawPublicKeyHex: String,
        requiredEventBit: UInt8 = 0x04
    ) -> Result<ParsedQRToken, TokenVerificationError> {
        switch parse(token: token) {
        case .failure(let error):
            return .failure(error)
        case .success(let parsed):
            guard parsed.version == "WW2" else {
                return .failure(.unsupportedVersion)
            }
            if (parsed.eventBitmask & requiredEventBit) == 0 {
                return .failure(.unauthorizedEvent)
            }

            guard let pubData = dataFromHex(rawPublicKeyHex),
                  let publicKey = try? P256.Signing.PublicKey(x963Representation: pubData) else {
                return .failure(.invalidKey)
            }
            return verify(parsed: parsed, publicKey: publicKey)
        }
    }

    /// DER/SPKI variant used by the signed Wedding Day manifest. This matches Android's key contract.
    public static func verifyAsymmetric(
        token: String,
        publicKeyDerBase64: String,
        requiredEventBit: UInt8 = 0x04
    ) -> Result<ParsedQRToken, TokenVerificationError> {
        switch parse(token: token) {
        case .failure(let error):
            return .failure(error)
        case .success(let parsed):
            guard parsed.version == "WW2" else {
                return .failure(.unsupportedVersion)
            }
            if (parsed.eventBitmask & requiredEventBit) == 0 {
                return .failure(.unauthorizedEvent)
            }
            guard let der = Data(base64Encoded: publicKeyDerBase64),
                  let publicKey = try? P256.Signing.PublicKey(derRepresentation: der) else {
                return .failure(.invalidKey)
            }
            return verify(parsed: parsed, publicKey: publicKey)
        }
    }

    /// Verifies a raw IEEE-P1363 P-256 signature over an arbitrary canonical payload.
    /// Used to authenticate the root-signed offline manifest before it is cached.
    public static func verifyP1363(
        payload: String,
        signatureHex: String,
        publicKeyDerBase64: String
    ) -> Bool {
        guard let der = Data(base64Encoded: publicKeyDerBase64),
              let publicKey = try? P256.Signing.PublicKey(derRepresentation: der),
              let signatureData = dataFromHex(signatureHex),
              let signature = try? P256.Signing.ECDSASignature(rawRepresentation: signatureData) else {
            return false
        }
        return publicKey.isValidSignature(signature, for: Data(payload.utf8))
    }

    private static func verify(
        parsed: ParsedQRToken,
        publicKey: P256.Signing.PublicKey
    ) -> Result<ParsedQRToken, TokenVerificationError> {
        let payload = "\(parsed.version).\(parsed.weddingShortId).\(parsed.passSerial).\(String(format: "%02x", parsed.eventBitmask)).\(parsed.nonce)"
        guard let sigData = dataFromHex(parsed.signature),
              let ecdsaSignature = try? P256.Signing.ECDSASignature(rawRepresentation: sigData) else {
            return .failure(.signatureMismatch)
        }

        if publicKey.isValidSignature(ecdsaSignature, for: Data(payload.utf8)) {
            return .success(parsed)
        } else {
            return .failure(.signatureMismatch)
        }
    }

    /// Legacy symmetric HMAC-SHA256 verifier (WW1)
    public static func verify(token: String, secretKey: String, requiredEventBit: UInt8 = 0x04) -> Result<ParsedQRToken, TokenVerificationError> {
        switch parse(token: token) {
        case .failure(let error):
            return .failure(error)
        case .success(let parsed):
            if (parsed.eventBitmask & requiredEventBit) == 0 {
                return .failure(.unauthorizedEvent)
            }

            let payloadPrefix = "\(parsed.version).\(parsed.weddingShortId).\(parsed.passSerial).\(String(format: "%02x", parsed.eventBitmask)).\(parsed.nonce)"
            let key = SymmetricKey(data: Data(secretKey.utf8))
            let signatureData = HMAC<SHA256>.authenticationCode(for: Data(payloadPrefix.utf8), using: key)
            let computedHex = signatureData.map { String(format: "%02x", $0) }.joined()
            let expected16Bytes = String(computedHex.prefix(32))

            if parsed.signature.lowercased() == expected16Bytes.lowercased() {
                return .success(parsed)
            } else {
                return .failure(.signatureMismatch)
            }
        }
    }

    private static func dataFromHex(_ hex: String) -> Data? {
        let clean = hex.trimmingCharacters(in: .whitespacesAndNewlines)
        guard clean.count % 2 == 0 else { return nil }
        var data = Data(capacity: clean.count / 2)
        var index = clean.startIndex
        for _ in 0..<(clean.count / 2) {
            let nextIndex = clean.index(index, offsetBy: 2)
            guard let byte = UInt8(clean[index..<nextIndex], radix: 16) else { return nil }
            data.append(byte)
            index = nextIndex
        }
        return data
    }
}

File: apps/android/app/src/main/java/pro/wewed/app/services/TokenVerifier.kt
package pro.wewed.app.services

import java.io.ByteArrayOutputStream
import java.security.KeyFactory
import java.security.PublicKey
import java.security.Signature
import java.security.spec.X509EncodedKeySpec
import java.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

data class ParsedQRToken(
    val version: String,
    val weddingShortId: String,
    val passSerial: String,
    val eventBitmask: Int,
    val nonce: String,
    val signature: String
)

sealed class TokenVerificationResult {
    data class Success(val token: ParsedQRToken) : TokenVerificationResult()
    data class Failure(val reason: String) : TokenVerificationResult()
}

object TokenVerifier {
    fun parse(token: String): TokenVerificationResult {
        val parts = token.trim().split(".")
        if (parts.size != 6) {
            return TokenVerificationResult.Failure("Invalid format: expected 6 parts, got ${parts.size}")
        }
        val version = parts[0]
        if (version != "WW1" && version != "WW2") {
            return TokenVerificationResult.Failure("Unsupported version: $version")
        }
        val weddingShortId = parts[1]
        val passSerial = parts[2]
        val eventBitmask = parts[3].toIntOrNull(16)
            ?: return TokenVerificationResult.Failure("Invalid event bitmask")
        val nonce = parts[4]
        val signature = parts[5]
        if (weddingShortId.isBlank() || passSerial.isBlank() || nonce.isBlank()) {
            return TokenVerificationResult.Failure("Invalid token identifiers")
        }

        return TokenVerificationResult.Success(
            ParsedQRToken(
                version = version,
                weddingShortId = weddingShortId,
                passSerial = passSerial,
                eventBitmask = eventBitmask,
                nonce = nonce,
                signature = signature
            )
        )
    }

    /**
     * Asymmetric ECDSA (NIST P-256 / SHA-256) signature verification (WW2 Canonical Standard).
     * Scanning devices verify against a pre-cached public key and never possess a pass-signing private key.
     */
    fun verifyAsymmetric(
        token: String,
        publicKeyDerBase64: String,
        requiredEventBit: Int = 0x04
    ): TokenVerificationResult {
        val parseResult = parse(token)
        if (parseResult is TokenVerificationResult.Failure) return parseResult
        val parsed = (parseResult as TokenVerificationResult.Success).token
        if (parsed.version != "WW2") {
            return TokenVerificationResult.Failure("Asymmetric verification requires WW2.")
        }
        if ((parsed.eventBitmask and requiredEventBit) == 0) {
            return TokenVerificationResult.Failure("Unauthorized for this wedding event.")
        }

        return try {
            val publicKey = decodePublicKey(publicKeyDerBase64)
            val payload = "${parsed.version}.${parsed.weddingShortId}.${parsed.passSerial}.${String.format("%02x", parsed.eventBitmask)}.${parsed.nonce}"
            if (verifyP1363Bytes(payload, parsed.signature, publicKey)) {
                TokenVerificationResult.Success(parsed)
            } else {
                TokenVerificationResult.Failure("Signature mismatch.")
            }
        } catch (e: Exception) {
            TokenVerificationResult.Failure("Verification exception: ${e.message}")
        }
    }

    /** Verifies an IEEE-P1363 P-256/SHA-256 signature over an arbitrary canonical payload. */
    fun verifyP1363(
        payload: String,
        signatureHex: String,
        publicKeyDerBase64: String
    ): Boolean {
        return try {
            verifyP1363Bytes(payload, signatureHex, decodePublicKey(publicKeyDerBase64))
        } catch (_: Exception) {
            false
        }
    }

    /** Legacy symmetric HMAC-SHA256 verifier (WW1). */
    fun verify(token: String, secretKey: String, requiredEventBit: Int = 0x04): TokenVerificationResult {
        val parseResult = parse(token)
        if (parseResult is TokenVerificationResult.Failure) return parseResult
        val parsed = (parseResult as TokenVerificationResult.Success).token

        if ((parsed.eventBitmask and requiredEventBit) == 0) {
            return TokenVerificationResult.Failure("Unauthorized for this wedding event.")
        }

        val payloadPrefix = "${parsed.version}.${parsed.weddingShortId}.${parsed.passSerial}.${String.format("%02x", parsed.eventBitmask)}.${parsed.nonce}"
        val keySpec = SecretKeySpec(secretKey.toByteArray(Charsets.UTF_8), "HmacSHA256")
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(keySpec)
        val hmacBytes = mac.doFinal(payloadPrefix.toByteArray(Charsets.UTF_8))
        val computedHex = hmacBytes.joinToString("") { "%02x".format(it) }
        val expected16Bytes = computedHex.take(32)

        return if (parsed.signature.equals(expected16Bytes, ignoreCase = true)) {
            TokenVerificationResult.Success(parsed)
        } else {
            TokenVerificationResult.Failure("Signature mismatch.")
        }
    }

    private fun decodePublicKey(publicKeyDerBase64: String): PublicKey {
        val spkiBytes = Base64.getDecoder().decode(publicKeyDerBase64)
        return KeyFactory.getInstance("EC").generatePublic(X509EncodedKeySpec(spkiBytes))
    }

    private fun verifyP1363Bytes(payload: String, signatureHex: String, publicKey: PublicKey): Boolean {
        val sigBytes = hexToBytes(signatureHex)
        if (sigBytes.size != 64) return false
        val ecdsa = Signature.getInstance("SHA256withECDSA")
        ecdsa.initVerify(publicKey)
        ecdsa.update(payload.toByteArray(Charsets.UTF_8))
        return ecdsa.verify(p1363ToDer(sigBytes))
    }

    private fun p1363ToDer(p1363: ByteArray): ByteArray {
        require(p1363.size == 64) { "P-256 IEEE-P1363 signature must be 64 bytes" }
        val r = p1363.copyOfRange(0, 32)
        val s = p1363.copyOfRange(32, 64)

        val out = ByteArrayOutputStream()
        writeDerInteger(out, r)
        writeDerInteger(out, s)
        val body = out.toByteArray()

        val seq = ByteArrayOutputStream()
        seq.write(0x30)
        seq.write(body.size)
        seq.write(body)
        return seq.toByteArray()
    }

    private fun writeDerInteger(out: ByteArrayOutputStream, value: ByteArray) {
        var start = 0
        while (start < value.size - 1 && value[start].toInt() == 0) start++
        val pad = (value[start].toInt() and 0x80) != 0
        val len = value.size - start + if (pad) 1 else 0
        out.write(0x02)
        out.write(len)
        if (pad) out.write(0x00)
        out.write(value, start, value.size - start)
    }

    private fun hexToBytes(value: String): ByteArray {
        val s = value.trim()
        require(s.length % 2 == 0) { "Hex value must have an even length" }
        val data = ByteArray(s.length / 2)
        var i = 0
        while (i < s.length) {
            val high = Character.digit(s[i], 16)
            val low = Character.digit(s[i + 1], 16)
            require(high >= 0 && low >= 0) { "Invalid hex value" }
            data[i / 2] = ((high shl 4) + low).toByte()
            i += 2
        }
        return data
    }
}
